src/hooks.js

```diff
@@ -0,0 +1,251 @@
+	currentComponent = c;
+	currentIndex = 0;
+}
+
+/**
+ * Get a hook's state from the currentComponent
+ * @param {number} index The index of the hook to get
+ * @returns {import('./internal').HookState}
+ */
+function getHookState(index) {
+	if (options._hook) options._hook(currentComponent);
+	// Largely inspired by:
+	// * https://github.com/michael-klein/funcy.js/blob/f6be73468e6ec46b0ff5aa3cc4c9baf72a29025a/src/hooks/core_hooks.mjs
+	// * https://github.com/michael-klein/funcy.js/blob/650beaa58c43c33a74820a3c98b3c7079cf2e333/src/renderer.mjs
+	// Other implementations to look at:
+	// * https://codesandbox.io/s/mnox05qp8
+	const hooks =
+		currentComponent.__hooks ||
+		(currentComponent.__hooks = { _list: [], _pendingEffects: [] });
+
+	if (index >= hooks._list.length) {
+		hooks._list.push({});
+	}
+	return hooks._list[index];
+}
+
+/**
+ * @param {import('./index').StateUpdater<any>} initialState
+ */
+export function useState(initialState) {
+	return useReducer(invokeOrReturn, initialState);
+}
+
+/**
+ * @param {import('./index').Reducer<any, any>} reducer
+ * @param {import('./index').StateUpdater<any>} initialState
+ * @param {(initialState: any) => void} [init]
+ * @returns {[ any, (state: any) => void ]}
+ */
+export function useReducer(reducer, initialState, init) {
+	/** @type {import('./internal').ReducerHookState} */
+	const hookState = getHookState(currentIndex++);
+	if (!hookState._component) {
+		hookState._component = currentComponent;
+
+		hookState._value = [
+			!init ? invokeOrReturn(undefined, initialState) : init(initialState),
+
+			action => {
+				const nextValue = reducer(hookState._value[0], action);
+				if (hookState._value[0] !== nextValue) {
+					hookState._value[0] = nextValue;
+					enqueueRender(hookState._component);
+				}
+			}
+		];
+	}
+
+	return hookState._value;
+}
+
+/**
+ * @param {import('./internal').Effect} callback
+ * @param {any[]} args
+ */
+export function useEffect(callback, args) {
+	/** @type {import('./internal').EffectHookState} */
+	const state = getHookState(currentIndex++);
+	if (argsChanged(state._args, args)) {
+		state._value = callback;
+		state._args = args;
+
+		currentComponent.__hooks._pendingEffects.push(state);
+	}
+}
+
+/**
+ * @param {import('./internal').Effect} callback
+ * @param {any[]} args
+ */
+export function useLayoutEffect(callback, args) {
+	/** @type {import('./internal').EffectHookState} */
+	const state = getHookState(currentIndex++);
+	if (argsChanged(state._args, args)) {
+		state._value = callback;
+		state._args = args;
+
+		currentComponent._renderCallbacks.push(state);
+	}
+}
+
+export function useRef(initialValue) {
+	return useMemo(() => ({ current: initialValue }), []);
+}
+
+/**
+ * @param {object} ref
+ * @param {() => object} createHandle
+ * @param {any[]} args
+ */
+export function useImperativeHandle(ref, createHandle, args) {
+	useLayoutEffect(
+		() => {
+			if (typeof ref === 'function') ref(createHandle());
```

Comment:
Would it be possible to reuse the logic from here:
https://github.com/preactjs/preact/blob/master/src/diff/index.js#L387